src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -0,0 +1,154 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#ifndef __SIMPLE_DATAFLOW_CPU_THREAD_HH__
+#define __SIMPLE_DATAFLOW_CPU_THREAD_HH__
+
+#include <vector>
+
+#include "cpu/exec_context.hh"
+#include "cpu/flexcpu/simple_dataflow_cpu.hh"
+#include "cpu/simple_thread.hh"
+
+
+/**
+ * Instances of this class represent a logical "Core" within the
+ * SimpleDataflowCPU.
+ */
+class SDCPUThread
+{
+  protected:
+    // BEGIN Internal class definitions
+
+    class SDCPUExecContext : public ExecContext
+    {
+      protected:
+        SDCPUThread& backingState;
+        
+      public:
+        SDCPUExecContext(SDCPUThread& backing_state):
+            backingState(backing_state)
+        { }
+    };
+
+    class DynamicInst
+    {
+        StaticInst& instructionRef;
+        // TODO instruction tag
+        // TODO instruction results
+    };
+
+    // END Internal class definitions
+
+    SimpleDataflowCPU* _cpuRef;
+
+    /**
+     * Monotonically increasing count of the instructions executed within this
+     * thread thus far. Since it should be unique for every (committed)
+     * execution of an instruction, this should be freely usable as a tag for
+     * our dynamic instructions.
+     */
+    typedef uint64_t InstTag;
+    InstTag numInstsCommitted = 0;
+    InstTag numInstsSpeculated = 0;
+
+    /**
+     * The SimpleThread class
+     */
+    SimpleThread _committedState;
+
+    // BEGIN Speculative state
+
+    /**
+     * Instruction execution will pass through an ExecContext interface,
+     * therefore in most cases, it will be tied to speculative state, instead
+     * of the "real" state of the system.
+     */
+    SDCPUExecContext _isaInterface;
+
+    TheISA::PCState _speculativePC;
+    // TODO hold speculative state.
+    // TODO hold table of in-flight instructions.
+
+    // END Speculative state
+
+    // TODO make sure this function is capable of holding off on issue in the
+    // case where we've only partially read in a macroOp
+    void decodeInstruction(TheISA::MachInst inst);
+
+    /**
+     * This function has the job of adding an instruction to the TODO table,
+     * populating its fields for dependencies, state, and tag. During this
+     * stage, after the instruction has effectively been "scheduled"
+     * internally, we should call executeAvailableInstructions to ask the CPU
+     * to begin execution of any instructions whose dependencies are all
+     * available.
+     * 
+     * (Note: Assumes in-order issue.)
+     * 
+     * @param inst The instruction to issue.
+     */
+    void issueInstruction(const StaticInst& inst);
+
+    void executeAvailableInstructions();
```

Comment:
I think this function should be in the CPU. The CPU is what will choose from all of the available instructions, which may be from many different threads.

The CPU will look at the "TODO" table you talk about above and execute as many instructions as it can. We don't need to worry about the priority/fairness of this. Now that I'm thinking about it, we obviously should execute the oldest instructions (that are ready) first.